goldenmaster/apigear/mqtt/mqttcppclient.cpp

```diff
@@ -0,0 +1,578 @@
+    std::thread([this, disconnectRequested](){
+        // we need to destroy the client outside of its own calling thread
+        MQTTAsync_destroy(m_client.get());
+        m_client.reset();
+
+        // reconnect if the connection was not dropped intentionally by us
+        if(!disconnectRequested)
+        {
+            // we need to re-subscribe to all topics on re-connection
+            m_subscribedTopicsMutex.lock();
+            m_toBeSubscribedTopicsMutex.lock();
+            if(!m_subscribedTopics.empty())
+            {
+                m_subscribedTopics.insert(m_toBeSubscribedTopics.begin(),m_toBeSubscribedTopics.end());
+                m_toBeSubscribedTopics = m_subscribedTopics;
+                m_subscribedTopics.clear();
+            }
+            m_toBeSubscribedTopicsMutex.unlock();
+            m_subscribedTopicsMutex.unlock();
+
+            connectToHost("");
+        }
+    }).detach();
+}
+
+void Client::handleTextMessage(const Message& message)
+{
+    // std::clog << "new msg topic: " << topic << " mqtt_message: " << message << std::endl;
+    std::pair<std::multimap<Topic, ISink*, Topic>::iterator, std::multimap<Topic, ISink*, Topic>::iterator> topics = m_subscribedTopics.equal_range(message.topic);
+    // for (auto test: m_subscribedTopics){
+    //     std::cout << "topic " << test.first << " " << test.second << std::endl; 
+    // }
+    for (auto iter = topics.first; iter != topics.second; ++iter)
+    {
+        if(iter->second != nullptr)
+        {
+            if(message.topic.getTopicType() == Topic::TopicType::Signal)
+            {
+                iter->second->onSignal(message.topic, message.content);
+            }
+            else if(message.topic.getTopicType() == Topic::TopicType::Property)
+            {
+                iter->second->onPropertyChanged(message.topic, message.content);
+            }
+            else if(message.topic.getTopicType() == Topic::TopicType::Operation)
+            {
+                iter->second->onInvoke(message.topic, message.content, message.responseTopic, message.correlationData);
+            }
+        }
+        else
+        {
+            // call locally
+            const int randomId = std::stoi(message.correlationData);
+            InvokeReplyFunc func {};
+            m_invokeResultsMutex.lock();
+            if((m_invokeResults.find(randomId) != m_invokeResults.end()))
+            {
+                func = m_invokeResults[randomId];
+                m_invokeResults.erase(randomId);
+            }
+            m_invokeResultsMutex.unlock();
+            if(func) {
+                const InvokeReplyArg arg{nlohmann::json::parse(message.content)};
+                func(arg);
+            }
+        }
+    }
+}
+
+void Client::invokeRemote(const Topic& topic, const std::string& value, InvokeReplyFunc func)
+{
+    const std::string responseTopic = topic.getEncodedTopic() + "/" + m_clientID + "/result" ;
```

Comment:
allocating string each time for sending was sth that slowed the cpp olink impl noticeably
In some project I used to work in with mqtt we declared all topics as const strings(but we didn't had to construct them, they were just some unique strings not dependent on id), so here we could generate it to some maps? vectors of vectors of vectors? just to find it and  use const string&

Reply:
done